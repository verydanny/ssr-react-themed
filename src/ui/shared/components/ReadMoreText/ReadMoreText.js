import React, { PureComponent } from 'react'
import themed from 'react-themed-too'
import PropTypes from 'prop-types'
import autobind from 'autobind-decorator'
import cn from 'classnames'
import truncateString from 'lib/utils/truncateString'

export const BUTTON_TEXT_EXPAND = 'Read more +'
export const BUTTON_TEXT_COLLAPSE = 'Read less'

@themed(/^ReadMoreText/, { pure: true })
export default class ReadMoreText extends PureComponent {

  static propTypes = {
    count: PropTypes.number.isRequired,
    text: PropTypes.string.isRequired,
    theme: PropTypes.object,
    triggerCount: PropTypes.number,
  }

  static defaultProps = {
    theme: {},
  }

  constructor(props) {
    super(props)
    this.state = { collapsed: true }
  }

  get triggerCount() {
    const { count, triggerCount } = this.props
    return triggerCount || count
  }

  get buttonText() {
    return this.state.collapsed ? BUTTON_TEXT_EXPAND : BUTTON_TEXT_COLLAPSE
  }

  get text() {
    const { count, text } = this.props

    if (this.shouldTruncate() && this.state.collapsed) {
      // Need a span here to support the HTML entity
      return <span>{truncateString(text, count)}&hellip;</span>
    }

    return text
  }

  shouldTruncate() {
    const { text } = this.props
    return (text.length > this.triggerCount)
  }

  @autobind
  handleClick() {
    this.setState(oldState => ({
      collapsed: !oldState.collapsed,
    }))
  }

  render() {
    const { theme } = this.props
    const shouldTruncate = this.shouldTruncate()
    const classNames = cn(
      theme.ReadMoreText,
      shouldTruncate && this.state.collapsed && theme['ReadMoreText-collapsed']
    )

    return (
      <div className={classNames} data-tid="read-more">
        <div className={theme.ReadMoreText_Text} data-tid="read-more-text">
          {this.text}
        </div>

        { shouldTruncate &&
          <div className={theme.ReadMoreText_Controls}>
            <button onClick={this.handleClick}>
              {this.buttonText}
            </button>
          </div>
        }

      </div>
    )
  }
}
