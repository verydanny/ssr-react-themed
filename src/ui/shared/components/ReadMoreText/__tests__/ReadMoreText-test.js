import React from 'react'
import { shallow } from 'enzyme'
import { keyMirror } from '@rentpath/react-ui-utils'
import ReadMoreText, {
  BUTTON_TEXT_EXPAND,
  BUTTON_TEXT_COLLAPSE,
} from '../ReadMoreText'

const ELLIPSIS = 'â€¦'

const theme = keyMirror([
  'ReadMoreText',
  'ReadMoreText_Text',
  'ReadMoreText-collapsed',
])

const line = ' The quick brown fox'
const lines2 = line.repeat(2)
const lines3 = line.repeat(3)
const lines4 = line.repeat(4)
const lines5 = line.repeat(5)

const defaultProps = {
  text: lines5,
  count: lines3.length,
  triggerCount: lines4.length,
  theme,
}

const factory = props =>
  shallow(<ReadMoreText.WrappedComponent {...defaultProps} {...props} />)

const validateNotTruncated = (wrapper, text) => {
  it('contains the original text', () => {
    expect(wrapper.text()).toBe(text)
  })
  it('does not have a read more button', () => {
    expect(wrapper.find('button').exists()).toBe(false)
  })
}

describe('<ReadMoreText />', () => {
  describe('text that should not truncate', () => {
    describe('text is smaller than the count', () => {
      const wrapper = factory({ text: lines2 })
      validateNotTruncated(wrapper, lines2)
    })
    describe('text is bigger than count but smaller than triggerCount', () => {
      const text = `${lines3} plus more`
      const wrapper = factory({ text })
      validateNotTruncated(wrapper, text)
    })
    describe('text is equal to the triggerCount', () => {
      const wrapper = factory({ text: lines4 })
      validateNotTruncated(wrapper, lines4)
      it('renders', () => {
        expect(wrapper.html()).toMatchSnapshot()
      })
    })
  })

  describe('text that should truncate', () => {
    const trunc = `${lines3}${ELLIPSIS}`
    describe('text is just over the triggerCount', () => {
      const text = `${lines4}+`
      const wrapper = factory({ text })
      it('contains truncated text', () => {
        expect(wrapper.find('.ReadMoreText_Text').text()).toBe(trunc)
      })
      it('has a read more button', () => {
        expect(wrapper.find('button').text()).toBe(BUTTON_TEXT_EXPAND)
      })
    })
    describe('text is well over the triggerCount', () => {
      const wrapper = factory({ text: lines5 })
      it('renders', () => {
        expect(wrapper.html()).toMatchSnapshot()
      })
      it('contains truncated text', () => {
        expect(wrapper.find('.ReadMoreText_Text').text()).toBe(trunc)
      })
      it('has a read more button', () => {
        expect(wrapper.find('button').text()).toBe(BUTTON_TEXT_EXPAND)
      })
    })
    describe('expanding and collapsing the text', () => {
      const wrapper = factory({ text: lines5 })
      wrapper.find('button').simulate('click')
      it('renders', () => {
        expect(wrapper.html()).toMatchSnapshot()
      })
      it('changes the text of the button', () => {
        expect(wrapper.find('button').text()).toBe(BUTTON_TEXT_COLLAPSE)
      })
      it('expands the text', () => {
        expect(wrapper.find('.ReadMoreText_Text').text()).toBe(lines5)
      })
      it('collapses the text again', () => {
        wrapper.find('button').simulate('click')
        expect(wrapper.find('button').text()).toBe(BUTTON_TEXT_EXPAND)
        expect(wrapper.find('.ReadMoreText_Text').text()).toBe(trunc)
      })
    })
  })
})
