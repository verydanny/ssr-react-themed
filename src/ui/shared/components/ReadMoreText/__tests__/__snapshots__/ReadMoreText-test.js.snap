// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ReadMoreText /> text that should not truncate text is equal to the triggerCount renders 1`] = `

<div class="ReadMoreText"
     data-tid="read-more"
>
  <div class="ReadMoreText_Text"
       data-tid="read-more-text"
  >
    The quick brown fox The quick brown fox The quick brown fox The quick brown fox
  </div>
</div>

`;

exports[`<ReadMoreText /> text that should truncate expanding and collapsing the text renders 1`] = `

<div class="ReadMoreText"
     data-tid="read-more"
>
  <div class="ReadMoreText_Text"
       data-tid="read-more-text"
  >
    The quick brown fox The quick brown fox The quick brown fox The quick brown fox The quick brown fox
  </div>
  <div>
    <button>
      Read less
    </button>
  </div>
</div>

`;

exports[`<ReadMoreText /> text that should truncate text is well over the triggerCount renders 1`] = `

<div class="ReadMoreText ReadMoreText-collapsed"
     data-tid="read-more"
>
  <div class="ReadMoreText_Text"
       data-tid="read-more-text"
  >
    <span>
      The quick brown fox The quick brown fox The quick brown foxâ€¦
    </span>
  </div>
  <div>
    <button>
      Read more +
    </button>
  </div>
</div>

`;
